### PR TITLE
Fix derivative step bug for funnels

### PR DESCRIPTION
**Pull request by EDsCODE**
## Changes

*Please describe.*  
- duplicate events are handled by shifting the window by 1 row when a step in the funnel matches the previous step. This ensures that the succeeding step in a funnel doesn't trivially consider the previous step as a conversion for the succeeding step
- This needs to be handled for events that might be generalized yet similar. For example, if step 1 is a $pageview with some properties and step 2 is just a $pageview, the query will commit the same error as described above with duplicated events. It will consider step 2 as being fulfilled trivially by step 1 since their both pageviews and we don't do the shift by 1 pattern when the succeding step is "similar" but not equal
- added test
*If this affects the frontend, include screenshots.*  

## Checklist

- [ ] All querysets/queries filter by Organization, by Team, and by User
- [ ] Django backend tests
- [ ] Jest frontend tests
- [ ] Cypress end-to-end tests
- [ ] Migrations are safe to run at scale (e.g. PostHog Cloud) – present proof if not obvious
- [ ] New/changed UI is decent on smartphones (viewport width around 360px)

### CODE REVIEW

**Comment by macobo** [comment]
Is there a relevant sentry issue?

File: posthog/models/entity.py [hunk 1]
```diff
@@ -69,6 +69,19 @@ def equals(self, other) -> bool:
...
 
+    """ Checks if this entity is a generalized version of other"""
+
+    def includes(self, other) -> bool:
+        self_properties = sorted([str(prop) for prop in self.properties])
+        other_properties = sorted([str(prop) for prop in other.properties])
+
+        num_matched_props = sum([1 for x, y in zip(self_properties, other_properties) if x == y])
+
+        if self.id == other.id and num_matched_props == len(self_properties):
```
Comment (macobo) [inline posthog/models/entity.py:80]: Proposal:

```python
return self.id == other.id and all(str(self_prop) == str(other_prop) for self_prop, other_prop in zip(self.properties, other.properties)
```

This simplifies this code significantly and is more idiomatic python

File: posthog/models/entity.py [hunk 2]
```diff
@@ -69,6 +69,19 @@ def equals(self, other) -> bool:
 
         return True
 
+    """ Checks if this entity is a generalized version of other"""
```
Comment (macobo) [inline posthog/models/entity.py:72]: Docstrings are added mostly _in_ the function, e.g. how it is above - should they be here?

Comment (macobo) [reply posthog/models/entity.py:72]: Q: What does generalized mean? 

If this means that it's equal in every other way than type, should we simplify equals to `return self.type == other.type and self.includes(other)` so the distiction becomes obvious?

**Review by macobo — CHANGES_REQUESTED** [review]

File: posthog/models/entity.py (hunk 2 above)
Comment (neilkakkar) [reply posthog/models/entity.py:72]: Judging from the code, I think it's supposed to mean: type is the same, and properties of one are subset of the other.

File: posthog/models/entity.py (hunk 1 above)
Comment (macobo) [reply posthog/models/entity.py]: Note: I might have misunderstood this, a length check might be needed as well (len(self.properties) >= len(other.properties

Comment (EDsCODE) [reply posthog/models/entity.py]: i'm changing it back. I forgot what i was checking at first, but we're not checking for equality here. We're just checking that the calling entity is a subset of the called on entity

File: posthog/models/entity.py [hunk 3]
```diff
@@ -69,6 +69,16 @@ def equals(self, other) -> bool:
 
         return True
 
+    def includes(self, other) -> bool:
+        """ Checks if this entity is a generalized version of other. The ids match and the properties of (this) is a subset of the properties of (other)"""
```
Comment (neilkakkar) [inline posthog/models/entity.py:73]: ok, starting to get confusing.

We want to go `1 PRECEDING` when the current entity is a superset of previous, right?

That implies checking `entities[i].is_superset(entities[i-1])`.

This does the reverse right now, doesn't it? 👀 

(Also, proposal on changing `includes` to `is_superset`/ `is_subset` ?

Comment (EDsCODE) [reply posthog/models/entity.py:73]: for it to be a superset, it has less properties than (other) so it has a `subset` of (other)'s properties.

`self._filter.entities[i].includes(self._filter.entities[i - 1])` is being called in base.py which should be correct

**Review by neilkakkar — APPROVED** [review]

File: posthog/models/entity.py (hunk 3 above)
Comment (neilkakkar) [reply posthog/models/entity.py]: Yep, looks good now! Thanks for this change! :)

`includes` == `is_superset` was what seemed confusing to me. ( I get now, in retrospect, includes implied properties, not the entity)

**Pull request merged**